### PR TITLE
clientv3/integration: add TestKVCompact

### PR DESCRIPTION
**Pull request by gyuho**
Finishing https://github.com/coreos/etcd/issues/4338.

### CODE REVIEW

**Comment by heyitsanthony** [comment]
lgtm-- should we also test that compact closes a watcher, though?

**Comment by gyuho** [comment]
Sure. Will add it too. Thanks.

On Thu, Feb 4, 2016, 08:30 Anthony Romano <notifications@github.com> wrote:

> lgtm-- should we also test that compact closes a watcher, though?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/coreos/etcd/pull/4413#issuecomment-179931509>.
>
-- 
Sincerely,
Gyu-Ho Lee

**Comment by xiang90** [comment]
> lgtm-- should we also test that compact closes a watcher, though?

hmm... This is testing the client actually. So we should not test the server behavior of closing the watchers here. Right?

File: clientv3/integration/kv_test.go [hunk 1]
```diff
@@ -251,3 +253,31 @@ func TestKVDelete(t *testing.T) {
...
+		if _, err := kv.Put("foo", "", lease.NoLease); err != nil {
+			t.Fatalf("couldn't put 'foo' (%v)", err)
+		}
+	}
+
+	if err := kv.Compact(2); err != nil {
+		t.Fatalf("couldn't compact kv space (%v)", err)
+	}
+	err := kv.Compact(2)
+	if err == nil || !strings.Contains(err.Error(), storage.ErrCompacted.Error()) {
```
Comment (xiang90) [inline clientv3/integration/kv_test.go]: can you please add these types to v3rpc/errors.go and add the error handling to togRPCError func?

Comment (xiang90) [inline clientv3/integration/kv_test.go]: then we can do err != v3rpc.ErrCompacted.

**Comment by heyitsanthony** [comment]
It would test the logic in the client responsible for closing an outdated watcher channel on a compact.

File: clientv3/integration/kv_test.go (hunk 1 above)
Comment (gyuho) [inline clientv3/integration/kv_test.go]: Yeah. Will add. Thanks.

On Thu, Feb 4, 2016, 09:17 Xiang Li <notifications@github.com> wrote:

> In clientv3/integration/kv_test.go
> <https://github.com/coreos/etcd/pull/4413#discussion_r51904815>:
>
> > +	clus := integration.NewClusterV3(t, &integration.ClusterConfig{Size: 3})
> > +	defer clus.Terminate(t)
> > +
> > +	kv := clientv3.NewKV(clus.RandClient())
> > +
> > +	for i := 0; i < 3; i++ {
> > +		if _, err := kv.Put("foo", "", lease.NoLease); err != nil {
> > +			t.Fatalf("couldn't put 'foo' (%v)", err)
> > +		}
> > +	}
> > +
> > +	if err := kv.Compact(2); err != nil {
> > +		t.Fatalf("couldn't compact kv space (%v)", err)
> > +	}
> > +	err := kv.Compact(2)
> > +	if err == nil || !strings.Contains(err.Error(), storage.ErrCompacted.Error()) {
>
> then we can do err != v3rpc.ErrCompacted.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/coreos/etcd/pull/4413/files#r51904815>.
>
-- 
Sincerely,
Gyu-Ho Lee

**Comment by xiang90** [comment]
@heyitsanthony OK. You are right.

**Comment by gyuho** [comment]
/cc @heyitsanthony @xiang90 

1. Checks if watch stream on compacted revision gets canceled
2. Use v3rpc errors for `ErrCompacted`

PTAL. 

Thanks.

File: clientv3/integration/kv_test.go [hunk 2]
```diff
@@ -251,3 +253,46 @@ func TestKVDelete(t *testing.T) {
...
+
+	wc := clientv3.NewWatcher(clus.RandClient())
+	defer wc.Close()
+	wchan := wc.Watch(context.TODO(), "foo", 1)
+
+	emptyResp := clientv3.WatchResponse{Header: pb.ResponseHeader{}, Events: nil}
+	wr, ok := <-wchan
+	empty := reflect.DeepEqual(emptyResp, wr)
+	if !empty {
+		t.Fatalf("expected empty response from closed channel, got %+v", wr)
```
Comment (xiang90) [inline clientv3/integration/kv_test.go]: we usually print out what we got first, then print out what we want.

**Comment by gyuho** [comment]
@xiang90 Just addressed. PTAL. Thanks.

File: clientv3/integration/kv_test.go [hunk 3]
```diff
@@ -251,3 +253,45 @@ func TestKVDelete(t *testing.T) {
...
+	if err == nil || err != v3rpc.ErrCompacted {
+		t.Fatalf("error got %v, want %v", err, v3rpc.ErrFutureRev)
+	}
+
+	wc := clientv3.NewWatcher(clus.RandClient())
+	defer wc.Close()
+	wchan := wc.Watch(context.TODO(), "foo", 3)
+
+	wr, ok := <-wchan
+	isEmpty := reflect.DeepEqual(clientv3.WatchResponse{Header: pb.ResponseHeader{}, Events: nil}, wr)
```
Comment (xiang90) [inline clientv3/integration/kv_test.go]: remove the empty checking since this ensures by go runtime that when a chan is closed it returns empty response.

**Comment by xiang90** [comment]
LGTM

**Comment by gyuho** [comment]
Just fixed. Will merge after green lights. Thanks!

**Comment by gyuho** [comment]
Semaphore is not responding... I ran integration test on Google cloud machine and confirmed that all tests pass.

**Pull request merged**